[FEAT] 채팅방 진입점 추가

kimdevv opened the pull request:
## Issue Number
closed #977 

## As-Is
- 예약 승인 후 만들어진 채팅방에 입장할 수 있는 방법이 없었습니다.

## To-Be
- 멘토가 자신이 개설한 멘토링에 달린 예약 목록 조회 시, 채팅방이 있는 경우 채팅방 id와 채팅 상태를 함께 반환합니다.
- 멘티가 자신이 신청한 예약 목록 조회 시, 채팅방이 있는 경우 채팅방 id도 함께 조회합니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
지금 `ChatRoom`과 `Reservation`이 연관 관계를 맺고 있지 않아서 예약 목록 조회 시 N+1 문제가 발생합니다. 빠르게 구현하기 위해 일단은 이렇게 구현했습니다. 추후 DTO 프로젝션이나 QueryDSL 도입이 필요할 것 같습니다. 그 외의 방법이 있다면 커멘트 달아주세요.

kimdevv committed:
feat: 멘토링 예약 조회 시 채팅방 정보 연동 및 관련 테스트 추가

- MentorMentoringReservationResponse에 chatRoomId, chatStatus 추가
- ParticipatedReservationResponse에 chatRoomId 필드 추가
- ReservationService에서 예약 데이터 조회 시 ChatRoom 정보 연동 로직 추가
- ChatRoomService에 findByReservationId 메서드 추가
- ChatRoomRepository에 예약 ID로 채팅방 조회 기능 추가
- 관련 테스트 코드 보완 및 새로운 로직에 맞게 수정
- FixtureUtil에 테스트용 ChatRoom 생성 메서드 추가

jye326 commented on backend/fittoring/src/main/java/fittoring/application/mentoring/service/dto/MentorMentoringReservationResponse.java:
```diff
@@ -10,14 +13,16 @@ public record MentorMentoringReservationResponse(
         String content,
         String status,
+        Long chatRoomId,
+        ChatStatus chatStatus,
```
p3 : ChatStatus(채팅의 상태)인가요? ChatRoomStatus(채팅방의 상태)인가요?

kimdevv replied:
채팅의 상태입니다. ACTIVATE랑 DEACTIVATE 상태가 있어서 채팅방 상태로 봐도 무방할 것 같아요.

jye326 reviewed (approved):
질문 남기고 승인합니다~ 수고하셨어요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/application/mentoring/service/dto/MentorMentoringReservationResponse.java:
```diff
@@ -10,14 +13,16 @@ public record MentorMentoringReservationResponse(
 ) {
 
-    public static MentorMentoringReservationResponse of(Reservation reservation) {
+    public static MentorMentoringReservationResponse of(Reservation reservation, Optional<ChatRoom> chatRoom) {
```
p2: optional 매개변수로 null 상태를 함께 처리하게 되면 호출할 때에도 optional을 보내야 하고, 내부에서도 isEmpty() 검증을 수행하게 되어 불편한 것 같아요. [Optional을 반환값 이외의 용도로 쓰는 경우는 거의 없다](https://jaehun2841.github.io/2019/02/24/effective-java-item55/#isPresent-%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EC%A7%80-%EB%A7%90%EC%9E%90)고 해요.

메서드 오버로딩을 통해서 처리하는건 어떠신가요? 

논리상으로도 `ofApprovedOrRejected`에서 optional 여부로 채팅룸 정보를 반환하는 것보다 예약 상태를 먼저 체크하고 채팅룸 정보가 있을 때 채팅룸 정보를 반환하는 것이 맞는 것 같아요!

```java
public static MentorMentoringReservationResponse of(Reservation reservation) {
    return of(reservation, null);
}

public static MentorMentoringReservationResponse of(Reservation reservation, ChatRoom chatRoom) {
    if (reservation.isPending()) {
        return ofPending(reservation);
    }
    return ofApprovedOrRejected(reservation, chatRoom);
}
```

kimdevv replied:
좋네요. 반영하겠습니다~

dye0p commented on backend/fittoring/src/main/java/fittoring/application/chat/service/ChatRoomService.java:
```diff
@@ -130,4 +131,8 @@ private MemberRole resolveChatMemberRole(Member member, Reservation reservation)
     }
+
+    public Optional<ChatRoom> findByReservationId(Long reservationId) {
+        return chatRoomRepository.findByReservationId(reservationId);
```
p2: 옵셔널로 반환하는 특별한 이유가 있을까요?

kimdevv replied:
이 메서드를 사용하는 부분에서 null을 다루어야 해서 Optional로 처리했습니다. 이 부분은 제거했습니다!

kimdevv replied:
해당 메서드를 사용하는 곳에서 null을 다루어야 해서 Optional로 했습니다. 그러나 조회 로직을 배치 조회로 수정하게 되어서 해당 메서드는 제거했습니다!

ppparkta reviewed (commented):
고생하셨습니다. 기능 상 문제는 없지만 optional 사용해주신 부분에 대해 코멘트 하나 남깁니다. 
확인부탁드려요!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/reservation/service/ReservationService.java:
```diff
@@ -121,15 +128,19 @@ public List<ParticipatedReservationResponse> findMemberReservations(Long memberI
         return rows.stream()
-                .map(r -> new ParticipatedReservationResponse(
+                .map(r -> {
+                    Optional<ChatRoom> chatRoom = chatRoomService.findByReservationId(r.getReservationId());
```
p2: 여기서 체이닝으로 `orElse(null)`을 해주면 어떨까요?

kimdevv replied:
필요한 부분이 chatRoom.getId()라서, 해당 부분에서 null을 먼저 반환해버리면 아래에서 NPE가 발생할 것 같습니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/reservation/service/ReservationService.java:
```diff
@@ -94,9 +97,13 @@ public List<MentorMentoringReservationResponse> getReservationsByMentor(Long men
+            MentorMentoringReservationResponse reservationResponse = MentorMentoringReservationResponse.of(reservation, chatRoom);
+            reservationResponses.add(reservationResponse);
+        }
+        return reservationResponses;
```
p1: 남겨주신 n + 1 문제에 대해서 reservationId를 LIst로 모아서 chatRoom을 batch(in 절 활용) 조회하는건 어떨까요? 
    `List<ChatRoom> findAllByReservationIdIn(List<Long> reservationIds);` -> chatRoomRepository 쿼리 메서드

말씀해주신 것 처럼 프로젝션을 이용해도 괜찮을 것 같네요~

kimdevv replied:
좋네요~ 배치 조회로 DB I/O를 1회만 수행하는 로직으로 바꿔보겠습니다!

dye0p reviewed (commented):
수고하셨어요~ 몇가지 코멘트 남겼으니 확인 부탁드립니다~

jye326 reviewed (approved):
승인합니다~

kimdevv committed:
feat: 채팅방 연동 로직 개선 및 테스트 보완

- ChatRoom 정보 조회 시 Batch Query 방식으로 변경하여 성능 최적화
- MentorMentoringReservationResponse 및 ParticipatedReservationResponse 관련 로직 수정
- ReservationService 내부 로직 간소화 및 가독성 개선
- ChatRoomService에 일괄 조회 메서드 추가 및 기존 로직 수정
- 테스트 코드에 새로운 로직 적용 및 ChatRoom 관련 시나리오 추가

kimdevv committed:
merge commit

kimdevv committed:
test: 테스트 선언문 수정

ppparkta reviewed (approved):
고생하셨습니다. 승인합니다!
